### PR TITLE
feature: Batch Writer

### LINKED ISSUE #6: Batch Writer
Batch Writer basic flow:
1) accept operations being delivered via Add method
2) 'cut' configurable number of operations into batch file 
3) store batch file into CAS (content addressable storage)
4) create Merkle tree from batch operations
4) create an anchor file based on batch file address and Merkle tree root
5) store anchor file into CAS
6) writes the address of anchor file to the underlying ledger

### PR DESCRIPTION
**Pull request by sandrask**
Batch Writer batches pending (Create, Update, Delete and Recover) operations and anchors them on the blockchain at a periodic interval

Basic flow:

- accept operations being delivered via Add method
- 'cut' configurable number of operations into batch file
- store batch file into CAS (content addressable storage)
- create Merkle tree from batch operations
- create an anchor file based on batch file address and Merkle tree root
- store anchor file into CAS
- writes the address of anchor file to the underlying ledger

Closes #6

Signed-off-by: Sandra Vrtikapa <sandra.vrtikapa@securekey.com>

### CODE REVIEW

File: pkg/api/batch.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package api
```
Comment (troyronda) [inline pkg/api/batch.go]: Generally interfaces should be defined in the packages that use them. Is there a reason for this separate API package?

File: pkg/utils/hash.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	return multihash.Encode(hash, uint64(multihashCode))
+
+}
+
+// GetHash will return hash based on specified multihash code
+func GetHash(multihashCode uint) (h hash.Hash, err error) {
+
+	switch multihashCode {
+	case 18:
```
Comment (rolsonquadras) [inline pkg/utils/hash.go]: A comment on what '18' means would be helpful for someone who is new to code.

File: pkg/api/protocol.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package api
+
+// Protocol defines parameter values.
+type Protocol struct {
+	/** The inclusive starting logical blockchain time that this protocol applies to. */
```
Comment (troyronda) [inline pkg/api/protocol.go]: Where are some description comments using "/**" vs "//"

File: pkg/api/batch.go (hunk 1 above)
Comment (sandrask) [reply pkg/api/batch.go]: I will move everything except for common structs and interfaces: blockchain, cas and protocol clients.

File: pkg/api/protocol.go (hunk 3 above)
Comment (sandrask) [reply pkg/api/protocol.go]: Nope, will change

File: pkg/utils/hash.go (hunk 2 above)
Comment (sandrask) [reply pkg/utils/hash.go]: I agree I will define meaningful constant

File: pkg/api/batch.go (hunk 1 above)
Comment (sandrask) [reply pkg/api/batch.go]: I have also removed blockchain and cas interfaces, only protocol stays in api.

File: pkg/api/protocol.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package api
```
Comment (troyronda) [inline pkg/api/protocol.go]: To be honest, this still seems like the protocol package

**Comment by troyronda** [comment]
@sandrask could you please squash the feat commits?

**Pull request merged**